spec/models/dynamic_field_group_spec.rb

```diff
@@ -0,0 +1,339 @@
+      end
+    end
+
+    context 'when creating a group with the same name as a sibling' do
+      context 'and the sibiling is a dynamic_field' do
+        let(:parent) { FactoryBot.create(:dynamic_field_group) }
+        let(:dynamic_field_group) { FactoryBot.build(:dynamic_field_group, :child, parent: parent) }
+
+        before do
+          FactoryBot.create(:dynamic_field, string_key: 'role', dynamic_field_group: parent)
+          parent.reload
+        end
+
+        it 'does not save' do
+          expect(dynamic_field_group.save).to be false
+        end
+
+        it 'returns correct error' do
+          dynamic_field_group.save
+          expect(dynamic_field_group.errors.full_messages).to include 'String key has already been taken'
+        end
+      end
+
+      context 'and the sibiling is a dynamic_field_group' do
+        let(:parent) { FactoryBot.create(:dynamic_field_category) }
+        let(:dynamic_field_group) { FactoryBot.build(:dynamic_field_group, parent: parent) }
+
+        before do
+          FactoryBot.create(:dynamic_field_group, string_key: 'name', parent: parent)
+          parent.reload
+        end
+
+        it 'does not save' do
+          expect(dynamic_field_group.save).to be false
+        end
+
+        it 'returns correct error' do
+          dynamic_field_group.save
+          expect(dynamic_field_group.errors.full_messages).to include 'String key has already been taken'
+        end
+      end
+    end
+
+    context 'when creating a group that shares the same string_key as a non-sibling group' do
+      let(:dynamic_field_group) { FactoryBot.create(:dynamic_field_group) }
+      let(:dynamic_field) { FactoryBot.build(:dynamic_field, string_key: 'name', dynamic_field_group: dynamic_field_group) }
+
+      it 'saves the object' do
+        expect(dynamic_field_group.save).to be true
+      end
+    end
+  end
+
+  describe '#parent' do
+    let(:dynamic_field_group) { FactoryBot.build(:dynamic_field_group, :child) }
+
+    before do
+      dynamic_field_group.parent = parent
+      dynamic_field_group.save
+      parent.reload
+    end
+
+    context 'when setting DynamicFieldCategory as parent' do
+      let(:parent) { FactoryBot.create(:dynamic_field_category) }
+
+      it 'adds parent' do
+        expect(dynamic_field_group.parent).to eq parent
+      end
+
+      it 'adds dynamic field group to dynamic field category' do
+        expect(parent.dynamic_field_groups).to match_array [dynamic_field_group]
+      end
+    end
+
+    context 'when setting DynamicFieldGroup as parent' do
+      let(:parent) { FactoryBot.create(:dynamic_field_group) }
+
+      it 'adds parent' do
+        expect(dynamic_field_group.parent).to eq parent
+      end
+
+      it 'adds dynamic field group to parent dynamic field group' do
+        expect(parent.dynamic_field_groups).to match_array [dynamic_field_group]
+      end
```

Comment:
We should add another test here that ensures that a dynamic_field_group cannot be added as its own parent (because that would lead to infinite loops elsewhere in the code as we traverse the field hierarchy).

Reply:
Good point.